pkg/services/syncer/types/custom.go

```diff
@@ -0,0 +1,164 @@
+/*
+Copyright © 2022 SUSE LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package types
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+
+	upgradev1 "github.com/rancher/system-upgrade-controller/pkg/apis/upgrade.cattle.io/v1"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	provv1 "github.com/rancher-sandbox/rancheros-operator/pkg/apis/rancheros.cattle.io/v1"
+	"github.com/rancher-sandbox/rancheros-operator/pkg/services/syncer/config"
+)
+
+type CustomSyncer struct {
+	upgradev1.ContainerSpec
+
+	MountPath  string `json:"mountPath"`
+	OutputFile string `json:"outputFile"`
+}
+
+func (j *CustomSyncer) toContainers(mount string) []corev1.Container {
+	return []corev1.Container{
+		{
+			VolumeMounts: []corev1.VolumeMount{{Name: "output",
+				MountPath: mount,
+			}},
+			Name:    "runner",
+			Image:   j.Image,
+			Command: j.Command,
+			Args:    j.Args,
+			EnvFrom: j.EnvFrom,
+			Env:     j.Env,
+		},
+	}
+}
+
+func (j *CustomSyncer) Sync(c config.Config, s provv1.ManagedOSVersionChannel) ([]provv1.ManagedOSVersion, error) {
+	logrus.Infof("Syncing '%s/%s'", s.Namespace, s.Name)
+
+	mountDir := j.MountPath
+	outFile := j.OutputFile
+	if mountDir == "" {
+		mountDir = "/data"
+	}
+	if outFile == "" {
```

Comment:
I am not sure I understand the purpose of it... The output file needs to be under the mountpoint right? I am wondering if we aren't missing a validator for that.

Is this output file being read somewhere?

Reply:
the approach of the operator is the following:

- It creates a pod with:
  - a segregated (no SA, etc) initcontainer which is the one specified in the `ManagedOSVersionChannel`. The container instead of outputing the json in stdout, it needs to push it into some file, you can override the default mount location in that way.
  - When the initcontainer ends, the json is expected to be in a file, which gets then printed to stdout by a container which keeps running until the operator deletes it back, once the data is captured.
  - 
This is because we can't rely on kubelet to retain logs of completed containers. My first approach was getting logs from a container that terminates, but that was flaky as I couldn't always access to its log.

Reply:
Oh I see this second reading pod scaped to me. Now it all makes sense :+1: